network/src/peers_registry.rs

```diff
@@ -88,259 +92,350 @@ type ProtocolConnec = (ProtocolId, UniqueConnec<(UnboundedSender<Bytes>, u8)>);
-    pub fn add_peer(&mut self, peer_id: PeerId, peer_connection: PeerConnection) -> PeerIndex {
-        self.peer_connections.or_insert(peer_id, peer_connection)
+    // registry a new peer
+    #[allow(clippy::needless_pass_by_value)]
+    fn new_peer(
+        &mut self,
+        peer_id: PeerId,
+        connected_addr: Multiaddr,
+        endpoint: Endpoint,
+    ) -> PeerIndex {
+        self.peer_store
+            .write()
+            .new_connected_peer(&peer_id, connected_addr.clone());
+        let mut peer = PeerConnection::new(connected_addr, endpoint);
+        peer.connected_time = Some(now_ms());
+        let peer_index = self.peers.or_insert(peer_id.clone(), peer);
+        debug!(target: "network", "allocate peer_index {} to peer {:?}", peer_index, peer_id);
+        peer_index
     }
 
     #[allow(clippy::needless_lifetimes)]
     #[inline]
     pub fn peers_iter<'a>(&'a self) -> impl Iterator<Item = (&'a PeerId, &'a PeerConnection)> {
-        self.peer_connections.iter()
+        self.peers.iter()
     }
 
     #[inline]
     pub fn get<'a>(&'a self, peer_id: &PeerId) -> Option<&'a PeerConnection> {
-        self.peer_connections.get(peer_id)
+        self.peers.get(peer_id)
     }
 
     #[inline]
     pub fn get_mut<'a>(&'a mut self, peer_id: &PeerId) -> Option<&'a mut PeerConnection> {
-        self.peer_connections.get_mut(peer_id)
+        self.peers.get_mut(peer_id)
     }
 
     pub fn connection_status(&self) -> ConnectionStatus {
         let mut total: u32 = 0;
-        let mut unreserved_incoming: u32 = 0;
-        let mut unreserved_outgoing: u32 = 0;
-        for (_, peer_connection) in self.peer_connections.iter() {
+        let mut unreserved_inbound: u32 = 0;
+        let mut unreserved_outbound: u32 = 0;
+        for (peer_id, peer_connection) in self.peers.iter() {
             total += 1;
-            if peer_connection.is_outgoing() {
-                unreserved_outgoing += 1;
+            if self.is_reserved(peer_id) {
+                continue;
+            }
+            if peer_connection.is_outbound() {
+                unreserved_outbound += 1;
             } else {
-                unreserved_incoming += 1;
+                unreserved_inbound += 1;
             }
         }
         ConnectionStatus {
             total,
-            unreserved_incoming,
-            unreserved_outgoing,
-            max_incoming: self.max_incoming,
-            max_outgoing: self.max_outgoing,
+            unreserved_inbound,
+            unreserved_outbound,
+            max_inbound: self.max_inbound,
+            max_outbound: self.max_outbound,
         }
     }
 
     #[inline]
     pub fn connected_peers_indexes<'a>(&'a self) -> impl Iterator<Item = PeerIndex> + 'a {
-        Box::new(
-            self.peer_connections
-                .peer_id_by_index
-                .iter()
-                .map(|(k, _v)| *k),
-        )
+        Box::new(self.peers.peer_id_by_index.iter().map(|(k, _v)| *k))
```

Comment:
with impl trait, you don't have to box it.
just return iterator directly
`self.peers.peer_id_by_index.iter().map(|(k, _v)| *k)`